### ISSUE TITLE
Support lambda when syntax

### ISSUE THREAD

**Issue by TimvdLippe** [mockito/mockito#643]
It would be awesome to support a syntax like this:
```java
Mockito.when(mock::method).isInvokedWith(foo,bar).thenAnswer((foo, bar) -> 5);
```

The benefits: true type safety. Currently we have an interim solution which helps a lot with Java 8, but it still [relies on casting](https://github.com/mockito/mockito/blob/master/src/main/java/org/mockito/internal/stubbing/answers/AnswerFunctionalInterfaces.java#L39).
Also the `when` syntax can be confusing compared to the `verify` syntax, since verify actually has the method call on the return value of `verify()` whereas `when` wants the method call inside its function argument.

This reads a lot more fluent imo and given the compiler warnings and similarity to `verify` could greatly benefit new users.

The biggest issue with this syntax is probably: how do we track the lambda reference? I have tried to come up with a proof of concept, but since `mock::method` just returns a `Function` without any points to what mock it was actually referring too, I could not extract the actual mock we should stub for. Maybe we can do some Bytecode magic there?

**Comment by szczepiq** [comment]
The benefits come at the cost of clarity and simplicity :)

However, it is definitely worth exploring. I'm wondering if Java8 has sufficient feature set that we can rethink the entire Mockito API :) After all, Mockito was built for java 6 and with the features available in java 6 :)

**Comment by bric3** [comment]
I would even say the feature of Java 5 as there wasn't so much in Java 6 ;)

**Comment by ChristianSchwarz** [comment]
@TimvdLippe I wonder how the signature of `isInvokedWith(foo,bar)` would look like, is it an Object-Vararg? If yes it is not possible to ensure a dev pass always the correct number of arguments and also type safty can't be guaranteed.

**Comment by TimvdLippe** [comment]
@szczepiq I would actually argue that this syntax is a lot more clearer than the current `when` syntax.

@ChristianSchwarz So since we pass a `Function` into `when`, we can match on the generics of the function. E.g. if we pass a function we 2 arguments, the definition of `when` should be:

```java
public <F, S, R> LambdaWhenMatcher3<F, S, R> when(BiFunction<F, S, R> function) {
  // Do stuff with the function and use generics to ensure LambdaWhenMatcher has the correct type
}
```

Then in `LambdaWhenMatcher3` we define `isInvokedWith`:

```java
public LambdaWhenReturn<F, S, R> isInvokedWith(F first, S second) {
  // Again check for the matchers etc...
}
```

And finally `thenAnswer` in `LambdaWhenReturn` could be:

```java
public void thenAnswer(BiFunction<F, S, R> function) {
}
```

This should be fully type-safe.

**Comment by ChristianSchwarz** [comment]
@TimvdLippe I see, thanks for the explanation.
Some questions:
 * How would this approch handle methods with an insane number of arguments? 
 * Are primitive argument types a problem here?
 * What about declared Exceptions ? Can this be solved by declaring `throws Thowable` on all `isInvokedWith(...)` variants?

Anyway I like the API cause it opens the possibility to have a single API to stubb void and returning methods in the same way.

**Comment by szczepiq** [comment]
>Anyway I like the API cause it opens the possibility to have a single API to stubb void and returning methods in the same way.

We already have this, doReturn :)

While we argue whether new syntax is clearer or not. Fact is that new syntax makes the stubbing declaration visually different from "normal" use of a mock. This may be a good or bad thing. I'm undecided :) We need more experimentation and feedback.

**Comment by ChristianSchwarz** [comment]
@szczepiq 
Yes from a pure technical view doXXX (action -> when -> call) is all you need but it doesn't read as good as the when->call->action syntax. Most possibly I would argue differently if we never had when(..).

----
Here is just an other idea to unifiy the stubbing API of void and returning methods (but it lacks typesafe answers):

 * `when(() -> mock.compareTo(any() ) ) .thenReturn(..);`
 * `when(() -> mock.voidMethod() ).thenThrow(..);`

The signature of `when(..)` would look simplified like this:

```
   Answer<T> when(Callable<T> call) 
   VoidAnswer when(Runnable call)
```

**Comment by szczepiq** [comment]
Very interesting! I would need to investigate more / think on it more to make a recommendation. This brainstorming is really valuable and can give us fresh perspective / ideas how to improve the API.

Currently, I want to push forward the design of #303

**Comment by TimvdLippe** [comment]
I think there are a couple of things we can improve with this syntax:

1. Handle void + returning methods the same way. Mockito internally handles based on the argument passed. No confusion between `doXXX` and `when`
2. The `ArgumentCaptor` vs `any` vs literal values. If you use any + literal, stuff breaks on run-time. This is bad, we can use the compiler with this syntax to warn users about this. Plus we can make the ArgumentCaptor typesafe, see #565 

A technical detail that might block this issue: how do we determine which method to mock if we only pass a lambda?

**Comment by szczepiq** [comment]
I'm not sure about 'isInvokedWith' syntax. It does not look very appealing to me. However, don't stop brainstorming and prototyping! ;)

**Comment by ChristianSchwarz** [comment]
> A technical detail that might block this issue: how do we determine which method to mock if we only pass a lambda?

I had them same problem while implementing a prototyp for my syntax proposal `when(()->{mock.voidMethod(any())}).then(..)`.  I tried this:
 1. invoke the passed lambda body  `()->{mock.voidMethod(any())}`
 1. intercept the call to the mock
 1. get the argument matchers 
 1. throw an exception if more than one mock method was invoked

I absolutly like the idea of one stubbing API to rule them all! Some mockito APIs become redundant over time (when/then vs. given/will, Rule vs Runner, ...).

**Comment by iirekm** [comment]
I've implemented an extension of Mockito which uses `when(() -> ...).then...` syntax to allow things like mocking `static` or `void` methods, it's at https://github.com/iirekm/misc/tree/master/ajmock

**Comment by TimvdLippe** [comment]
I have another prototype written down locally, but haven't finished it. It
follows a very similar pattern. Haven't figured out the tiniest details
yet, but it will probably follow along your pattern :)

Op wo 1 nov. 2017 om 18:50 schreef Irek Matysiewicz <
notifications@github.com>:

**Comment by ChristianSchwarz** [comment]
@iirekm 
Interesting to see that the syntax `when(() -> ...).thenXXX()` is that popular ;-). I wonder how static mocking works in your proposal, how do you intercept a static call?

**Comment by iirekm** [comment]
I intercept it with an AspectJ aspect, which delegates to designated Mockito mock, so for example `C.staticMethod(args...)` behind the scenes is converted to something like: `getMockFor(C.class, "staticMethod").invoke(args...)`
